allow chrono constants to be used on device (with cuda 10.{0-1} compilers)

sriramch opened the pull request:
- constexpr functions may be used on device when compiling with `--expt-relaxed-constexpr.` some of these functions end up using types that are also annotated with constexpr, yet can't be used on device with nvcc 10.0 and 10.1 compilers. the weekday and the month constexpr definitions are the types that are referenced in some of the fundamental functions in the std chrono library
- whilst the weekday and month types may be used in 10.2, using them inside a kernel produces a device sync. (runtime) failure
- this pr annotates those symbols explicitly with `__device__ const` to quiesce the compiler
- this will let us write more standard conforming code without crafting ugly workarounds or copying `std::chrono` code to workaround compiler issues

@trxcllnt please review.

sriramch committed:
- constexpr functions may be used on device when compiling with `--expt-relaxed-constexpr.`
  some of these functions end up using types that are also annotated with constexpr, yet
  can't be used on device with nvcc 10.0 and 10.1 compilers. the weekday and the month
  constexpr definitions are the types that are referenced in some of the fundamental functions
  in the std chrono library.
  this pr annotates those symbols explicitly with `__device__ const` to quiesce the compiler.

jrhemstad commented:
We should not be modifying our local fork of libcu++. If this is a change we want made, it needs to be done through the libcu++ team so they can review it.

sriramch commented:
i do not believe that this change is even applicable/required with libcu++. for one, the code is quite different from what is present here and what is present in libcu++.

the reason for this change is documented in the pr description. hence, until we move to libcu++, if we need to support 10.0 and 10.1 compilers (at least the latter), we would need this; else, we may have to come up with some workarounds and/or copy parts of std chrono code into cudf.

jrhemstad commented:
So you're saying this change isn't necessary when we move to libcu++ instead of the freestanding fork?

sriramch commented:
> So you're saying this change isn't necessary when we move to libcu++ instead of the freestanding fork?

i have verified that we do *not* need this with cuda11 (and use `cuda/std/chrono`) and a compiler that supports `c++17` (not gcc 5.4). i have not tried this with a standalone libcu++ (outside of what comes bundled with cuda11).

jrhemstad commented:
Can you include an example of when/where this change is necessary?

sriramch commented:
```
//#include <cuda/std/chrono>
#include <simt/chrono>
#include <iostream>

using duration_d = simt::std::chrono::duration<int32_t, simt::std::ratio<86400>>;
using duration_m = simt::std::chrono::duration<int32_t, simt::std::ratio<2629746>>;

template <class Duration>
using timestamp = simt::std::chrono::time_point<simt::std::chrono::system_clock, Duration>;

using timestamp_d = timestamp<duration_d>;

__global__ void round_up_current_date(int dt) {
    using namespace simt::std::chrono;

    const auto nd = floor<days>(time_point<simt::std::chrono::system_clock, seconds>{seconds(dt)});
    const year_month_day ymd(nd);
    const year_month_day_last ymld(ymd.year(), month_day_last(ymd.month()));

    printf("rounded date on kernel: %u/%u/%d\n", static_cast<unsigned>(ymld.month()), static_cast<unsigned>(ymld.day()), static_cast<int>(ymld.year()));
}

int main(){
    using namespace simt::std::chrono;

    // this block does not build on 10.0 and 10.1
#if 1
    auto dt = year{2020}/month{3}/day{31}; // this is ymd
    dt += duration_m{-1};
    if (!dt.ok()) {
      dt = dt.year()/dt.month()/last;
    }

    timestamp_d dys = sys_days{dt};
    std::time_t ttp = simt::std::chrono::system_clock::to_time_t(dys);
    std::cout << "time: " << std::ctime(&ttp) << std::endl;
#endif

    // this block does not build on 10.0 and 10.1
    // this block does not run on 10.2
#if 1
    round_up_current_date<<<1, 1>>>(time(NULL));
    auto err = cudaDeviceSynchronize();
    if (err != cudaSuccess) {
      std::cout << "device sync. error " << err << ": " << cudaGetErrorString(err) << std::endl;
      return -1;
    }
#endif

    return 0;
}
```

sriramch commented:
@jrhemstad @trxcllnt - is there something you want me to do with this? if not, can this be merged? 

i'm presently struggling to workaround the issues that this is presenting while using `std::chrono` and would like to use the standard as much as possible. moving to `libcu++` will also be easier later without having to remove those workarounds.

jrhemstad commented:
I'm not okay with merging this without having someone from the libcu++ team look at this. I don't know all the cases when and where `_LIBCPP_INLINE_CONSTEXPR` is being used to know if your change is valid. For one thing, it doesn't include `static` which means the linkage is going to be external, whereas the other host cases provide internal linkage.

sriramch commented:
can you provide me some contacts from `libcu++` to look at this (or copy them here)?

the `_LIBCPP_INLINE_CONSTEXPR` is used only in the chrono header and is used in the context of `weekday` and `month` global const definitions.

the `const` specifier makes the linkage internal. more information is [here](https://stackoverflow.com/questions/3709207/c-semantics-of-static-const-vs-const)

fwiw, i built cudf with this change and a `std::chrono` compliant code (similar to above) and it works.

jrhemstad commented:
> the `const` specifier makes the linkage internal. more information is [here](https://stackoverflow.com/questions/3709207/c-semantics-of-static-const-vs-const)

Ah yes, I forgot about that.

Does it not work to mark the global constants simply as `constexpr`?

sriramch commented:
> Does it not work to mark the global constants simply as `constexpr`?

no and that was the initial cause of this issue. even with `--expt-relaxed-constexpr` the 10.x compilers expect these types to be device consts; either that, or these objects should be copied in places where it is used as local variables.

who else do you think should review this? the `libcu++` implementation has evolved quite a bit from this though.

harrism commented:
I think we should decouple this from libcudf. We should file an nvbug against libcu++ with a standalone repro of the issue. @sriramch can you create a standalone (single .cu file) repro of the issue? I can help you with the nvbug over slack.

sriramch committed:
- workaround yet another nvcc bug
  - on 10.x const global object of a type can't be created with initializer syntax `{}` on
    the device when the constructor is `explicit`
  - either removing the `explicit` keyword or not defining a default constructor also works
  - i have tried to make the definition consistent with the other types for which objects
    are defined at global scope

sriramch commented:
@harrism the reproducer is in this pr. 

i have slacked you with some questions.
